src/Microsoft.Health.Fhir.Core/Features/Operations/ConvertData/ContainerRegistryTemplateProvider.cs

```diff
@@ -0,0 +1,161 @@
+            EnsureArg.IsNotNull(containerRegistryTokenProvider, nameof(containerRegistryTokenProvider));
+            EnsureArg.IsNotNull(convertDataConfig, nameof(convertDataConfig));
+            EnsureArg.IsNotNull(logger, nameof(logger));
+
+            _containerRegistryTokenProvider = containerRegistryTokenProvider;
+            _convertDataConfig = convertDataConfig.Value;
+            _logger = logger;
+
+            // Initialize cache and template collection provider factory
+            _cache = new MemoryCache(new MemoryCacheOptions
+            {
+                SizeLimit = _convertDataConfig.CacheSizeLimit,
+            });
+            _templateCollectionProviderFactory = new TemplateCollectionProviderFactory(_cache, Options.Create(_convertDataConfig.TemplateCollectionOptions));
+        }
+
+        /// <summary>
+        /// Fetch template collection from container registry or built-in archive.
+        /// </summary>
+        /// <param name="request">The data convert request which contains template reference.</param>
+        /// <param name="cancellationToken">Cancellation token to cancel the fetch operation.</param>
+        /// <returns>Template collection.</returns>
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(ConvertDataRequest request, CancellationToken cancellationToken)
+        {
+            // We have embedded a default template collection in the templatemanagement package.
+            // If the template collection is the default reference, we don't need to retrieve token.
+            var accessToken = string.Empty;
+            if (!request.IsDefaultTemplateReference)
+            {
+                _logger.LogInformation("Using a custom template collection for data conversion.");
+
+                async Task<string> TokenEntryFactory(ICacheEntry entry)
+                {
+                    var token = await _containerRegistryTokenProvider.GetTokenAsync(request.RegistryServer, cancellationToken);
+                    entry.Size = token.Length;
+                    entry.AbsoluteExpiration = GetTokenAbsoluteExpiration(token);
+                    return token;
+                }
+
+                accessToken = await _cache.GetOrCreateAsync(GetCacheKey(request.RegistryServer), TokenEntryFactory);
+            }
+            else
+            {
+                _logger.LogInformation("Using the default template collection for data conversion.");
+            }
+
+            try
+            {
+                var provider = _templateCollectionProviderFactory.CreateTemplateCollectionProvider(request.TemplateCollectionReference, accessToken);
+                return await provider.GetTemplateCollectionAsync(cancellationToken);
+            }
+            catch (ContainerRegistryAuthenticationException authEx)
+            {
+                // Remove token from cache when authentication failed.
+                _cache.Remove(GetCacheKey(request.RegistryServer));
```

Comment:
Is it possible for token to expire and if so, shouldn't we try to acquire new token rather than just fail?